File: CarHomeWidget/src/main/java/com/anod/car/home/prefs/lookandfeel/SkinPreviewFragment.java
package com.anod.car.home.prefs.lookandfeel;

import com.anod.car.home.R;
import com.anod.car.home.appwidget.WidgetViewBuilder;
import com.anod.car.home.model.ShortcutInfo;
import com.anod.car.home.model.WidgetShortcutsModel;
import com.anod.car.home.prefs.LookAndFeelActivity;
import com.anod.car.home.prefs.drag.ShortcutShadowBuilder;
import com.anod.car.home.utils.AppLog;

import android.app.Activity;
import android.content.ClipData;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RemoteViews;

import butterknife.ButterKnife;
import butterknife.InjectView;

public class SkinPreviewFragment extends Fragment
        implements LoaderManager.LoaderCallbacks<View>, View.OnLongClickListener {

    private static final String ARG_POSITION = "position";

    private int mPosition;

    private LookAndFeelActivity mActivity;

    @InjectView(R.id.container_preview)
    ViewGroup mContainer;

    private int mShotcutsCount;

    public static SkinPreviewFragment newInstance(int position) {
        SkinPreviewFragment f = new SkinPreviewFragment();

        Bundle args = new Bundle();
        args.putInt(ARG_POSITION, position);

        f.setArguments(args);

        return f;
    }


    @Override
    public void onResume() {
        super.onResume();
        getLoaderManager().initLoader(0, null, this).forceLoad();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mActivity.onPreviewStart(mPosition);
        getLoaderManager().initLoader(0, null, this);
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        mPosition = getArguments().getInt(ARG_POSITION);
        mActivity = (LookAndFeelActivity) activity;
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.skin_item, container, false);
        ButterKnife.inject(this, view);
        return view;
    }

    public static class ViewLoader extends AsyncTaskLoader<View> {

        private final LookAndFeelActivity mActivity;

        private final int mPosition;

        public ViewLoader(LookAndFeelActivity activity, int position) {
            super(activity);
            mActivity = activity;
            mPosition = position;
        }

        @Override
        public View loadInBackground() {
            WidgetViewBuilder builder = mActivity.createBuilder();
            builder.init();

            builder.setOverrideSkin(mActivity.getSkinItem(mPosition).value);
            RemoteViews rv = builder.build();

            return rv.apply(mActivity, null);
        }

    }

    @Override
    public Loader<View> onCreateLoader(int id, Bundle args) {
        return new ViewLoader(mActivity, mPosition);
    }

    @Override
    public void onLoadFinished(Loader<View> loader, View inflatedView) {
        mActivity.onPreviewCreated(mPosition);

        if (inflatedView.getParent() != null) {
//            View parent = (View) inflatedView.getParent();
//            mContainer.addView(parent);
            return;
        }

        if (mContainer.getChildCount() > 0) {
            mContainer.removeAllViews();
        }

        final WidgetShortcutsModel model = new WidgetShortcutsModel(getActivity(),
                mActivity.getAppWidgetId());
        model.init();
        mShotcutsCount = model.getCount();

        setupDragNDrop(inflatedView, model);

        mContainer.addView(inflatedView);
//        mContainer.invalidate();
    }

    @Override
    public void onDestroyView() {
        if (mContainer != null && mContainer.getChildCount() > 0) {
            for (int pos = 0; pos < mShotcutsCount; pos++) {
                int btnResId = WidgetViewBuilder.getBtnRes(pos);
                final ImageView dragButton = (ImageView) mContainer.findViewById(btnResId);
                if (dragButton == null) {
                    AppLog.e("Count: " + mShotcutsCount + ", pos: " + pos);
                    continue;
                }
                dragButton.setOnLongClickListener(null);
                dragButton.setOnDragListener(null);
            }
        }

        super.onDestroyView();
    }

    private void setupDragNDrop(View inflatedView, final WidgetShortcutsModel model) {

        for (int pos = 0; pos < mShotcutsCount; pos++) {
            int btnResId = WidgetViewBuilder.getBtnRes(pos);
            final ImageView btn = (ImageView) inflatedView.findViewById(btnResId);
            if (btn == null) {
                AppLog.e("Count: " + mShotcutsCount + ", pos: " + pos);
                continue;
            }
            ShortcutInfo shortcut = model.getShortcut(pos);

            initDragButton(pos, btn, shortcut != null);
        }
    }

    @Override
    public boolean onLongClick(View dragButton) {
        String dragData = "" + dragButton.getTag();
        ClipData data = ClipData.newPlainText(dragData, dragData);
        mActivity.onBeforeDragStart();
        dragButton.startDrag(data, new ShortcutShadowBuilder(dragButton), null, 0);
        return true;
    }


    private void initDragButton(final int cellId, ImageView dragButton,
            final boolean hasShortcut) {
        dragButton.setTag(String.valueOf(cellId));
        if (hasShortcut) {
            dragButton.setOnLongClickListener(this);
        }
        dragButton.setOnDragListener(mActivity.getDragListener());
    }

    @Override
    public void onLoaderReset(Loader<View> loader) {

    }

}

File: CarHomeWidget/src/main/java/com/anod/car/home/prefs/lookandfeel/SkinPreviewFragment.kt
package com.anod.car.home.prefs.lookandfeel

import android.app.Application
import android.content.ClipData
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.InflateException
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.lifecycle.*
import com.anod.car.home.R
import com.anod.car.home.appwidget.WidgetViewBuilder
import com.anod.car.home.model.WidgetShortcutsModel
import com.anod.car.home.prefs.LookAndFeelActivity
import com.anod.car.home.prefs.drag.ShortcutShadowBuilder
import info.anodsplace.framework.AppLog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SkinPreviewViewModel(application: Application): AndroidViewModel(application) {
    val view = MutableLiveData<View>()
    lateinit var builder: WidgetViewBuilder
    var overrideSkin: String? = null

    fun load() {
        viewModelScope.launch {
            val preview = renderPreview(builder, overrideSkin)
            view.value = preview
        }
    }

    private suspend fun renderPreview(builder: WidgetViewBuilder, overrideSkin: String?): View = withContext(Dispatchers.Default) {
        builder.init()
        builder.overrideSkin = overrideSkin
        val rv = builder.build()
        try {
            return@withContext rv.apply(getApplication(), null)
        } catch (e: InflateException) {
            AppLog.e("Cannot generate preview for $overrideSkin", e)
            return@withContext TextView(getApplication()).apply {
                text = "Cannot generate preview"
            }
        }
    }
}

class SkinPreviewFragment : Fragment(), View.OnLongClickListener {

    private var position: Int = 0

    private var lookAndFeelActivity: LookAndFeelActivity? = null

    private var viewGroup: ViewGroup? = null

    private var shortcutsCount: Int = 0

    private val viewModel: SkinPreviewViewModel by lazy { ViewModelProviders.of(this).get(SkinPreviewViewModel::class.java) }

    override fun onResume() {
        super.onResume()
        viewModel.load()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        lookAndFeelActivity!!.onPreviewStart(position)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        position = arguments!!.getInt(ARG_POSITION)
        lookAndFeelActivity = context as LookAndFeelActivity?
    }

    override fun onDetach() {
        super.onDetach()
        lookAndFeelActivity = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.skin_item, container, false)
        viewGroup = view.findViewById(R.id.container_preview)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.overrideSkin = lookAndFeelActivity!!.getSkinItem(position).value
        viewModel.builder = lookAndFeelActivity!!.createBuilder()
        viewModel.view.observe(this, Observer { inflatedView ->
            lookAndFeelActivity!!.onPreviewCreated(position)

            if (inflatedView.parent != null) {
                return@Observer
            }

            if (viewGroup!!.childCount > 0) {
                viewGroup!!.removeAllViews()
            }

            val model = WidgetShortcutsModel.init(activity!!, lookAndFeelActivity!!.appWidgetId)
            shortcutsCount = model.count

            setupDragNDrop(inflatedView, model)
            viewGroup!!.addView(inflatedView)
        })
    }

    fun refresh() {
        viewModel.load()
    }

    override fun onDestroyView() {
        if (viewGroup != null && viewGroup!!.childCount > 0) {
            for (pos in 0 until shortcutsCount) {
                val btnResId = WidgetViewBuilder.btnIds[pos]
                val dragButton = viewGroup!!.findViewById<View>(btnResId) as? ImageView
                if (dragButton == null) {
                    AppLog.e("Count: $shortcutsCount, pos: $pos")
                    continue
                }
                dragButton.setOnLongClickListener(null)
                dragButton.setOnDragListener(null)
            }
        }

        super.onDestroyView()
    }

    private fun setupDragNDrop(inflatedView: View, model: WidgetShortcutsModel) {

        for (pos in 0 until shortcutsCount) {
            val btnResId = WidgetViewBuilder.btnIds[pos]
            val btn = inflatedView.findViewById<View>(btnResId) as? ImageView
            if (btn == null) {
                AppLog.e("Count: $shortcutsCount, pos: $pos")
                continue
            }
            val shortcut = model.get(pos)

            initDragButton(pos, btn, shortcut != null)
        }
    }

    override fun onLongClick(dragButton: View): Boolean {
        val dragData = "" + dragButton.tag
        val data = ClipData.newPlainText(dragData, dragData)
        lookAndFeelActivity!!.onBeforeDragStart()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            dragButton.startDragAndDrop(data, ShortcutShadowBuilder(dragButton), null, 0)
        } else {
            dragButton.startDrag(data, ShortcutShadowBuilder(dragButton), null, 0)
        }
        return true
    }


    private fun initDragButton(cellId: Int, dragButton: ImageView,
                               hasShortcut: Boolean) {
        dragButton.tag = cellId.toString()
        if (hasShortcut) {
            dragButton.setOnLongClickListener(this)
        }
        dragButton.setOnDragListener(lookAndFeelActivity!!.dragListener)
    }

    companion object {
        private const val ARG_POSITION = "position"
        fun newInstance(position: Int) = SkinPreviewFragment().apply {
            arguments = bundleOf(ARG_POSITION to position)
        }
    }

}
